src/utils/ftexpression.rs

```diff
@@ -0,0 +1,194 @@
+            "=" => value == *thr,
+            "!=" => value != *thr,
+            _ => {
+                eprintln!("Invalid operator for {} function: {}", name, operator);
+                std::process::exit(1);
+            }
+        },
+        Threshold::Range(min, max) => match operator {
+            "=" => value >= *min && value < *max,
+            _ => {
+                eprintln!(
+                    "Invalid operator for {} function with range: {}",
+                    name, operator
+                );
+                std::process::exit(1);
+            }
+        },
+    }
+}
+
+fn len(value: i64, operator: &str, threshold: &Threshold) -> bool {
+    cmp("len", value as f64, operator, threshold)
+}
+
+fn qual(value: u8, operator: &str, threshold: &Threshold) -> bool {
+    cmp("qual", value as f64, operator, threshold)
+}
+
+pub struct ParsedExpr {
+    fn_name: String,
+    pub rng_name: String,
+    op: String, // <, <=, >, >=, etc
+    threshold: Threshold,
+}
+
+pub fn parse_filter(filter_orig: &str) -> ParsedExpr {
+    let mut filter = filter_orig.to_string();
+    filter.retain(|c| !c.is_whitespace());
+
+    let func_name_end = filter.find('(').unwrap_or(filter.len());
+    let func_name = filter[..func_name_end].trim().to_string();
+
+    let gnm_feat_start = filter.find('(').unwrap_or(filter.len()) + 1;
+    let gnm_feat_end = filter.find(')').unwrap_or(filter.len());
+    let gnm_feat = filter[gnm_feat_start..gnm_feat_end].to_string();
+    if !["msp", "nuc", "m6a", "5mC"].contains(&gnm_feat.as_str()) {
+        eprintln!("Invalid argument for len function: {}", gnm_feat);
+        std::process::exit(1);
+    }
+
+    let rest = &filter[gnm_feat_end + 1..].trim();
+
+    let operators = ["!=", ">=", "<=", ">", "<", "="];
+    let mut operator = "".to_string();
+    let mut threshold = None;
+    let mut range = None;
+
+    for &op in operators.iter() {
+        if let Some(pos) = rest.find(op) {
+            operator = op.to_string();
+            let threshold_str = rest[pos + op.len()..].trim();
+            if threshold_str.contains(':') {
+                let range_parts: Vec<&str> = threshold_str.split(':').collect();
+                if range_parts.len() == 2 {
+                    range = Some((
+                        range_parts[0].trim().parse::<f64>().unwrap(),
+                        range_parts[1].trim().parse::<f64>().unwrap(),
+                    ));
+                }
+            } else {
+                threshold = Some(threshold_str.parse::<f64>().unwrap());
+            }
+            break;
+        }
+    }
+
+    if let Some((_, _)) = range {
+        if operator != "=" {
+            eprintln!("Range thresholds can only be used with the '=' operator.");
+            std::process::exit(1);
+        }
+    }
+
+    let threshold_value = match range {
+        Some((min, max)) => Threshold::Range(min, max),
+        None => Threshold::Single(threshold.unwrap()),
+    };
```

Comment:
I think it is important to be able to apply multiple filters at once in a single command such that this function returns a Vec of `ParsedExp`, which can then be iteritivaly applied to every fiber-seq record.

Reply:
d2f3cbd307ffde66a7dd42458cc937020451ab3f